### PR TITLE
Use Two way algorithm to optimize ByteBufUtil.indexOf() method

### PR DESCRIPTION
**Pull request by skyguard1**
Motivation:

ByteBufUtil.indexOf can be inefficient for substring search on
ByteBuf, in terms of algorithm complexity (O(needle.readableBytes * haystack.readableBytes)), consider using the Two Way  algorithm to optimize the ByteBufUtil.indexOf() method

Modification:

Use the Two Way algorithm to optimize ByteBufUtil.indexOf() method.

Result:

The performance of the ByteBufUtil.indexOf() method is higher than the original implementation

### CODE REVIEW

**Comment by skyguard1** [comment]
Benchmark results (original implementation  of ByteBufUtil.indexOf):
```
Benchmark                (bufferType)        (input)   Mode  Cnt     Score     Error   Units
SearchBenchmark.indexOf          HEAP    RANDOM_256B  thrpt    5  1440.368 ±  41.740  ops/ms
SearchBenchmark.indexOf          HEAP     RANDOM_2KB  thrpt    5   183.511 ±   2.488  ops/ms
SearchBenchmark.indexOf          HEAP    PREDICTABLE  thrpt    5    42.856 ±   2.517  ops/ms
SearchBenchmark.indexOf          HEAP  UNPREDICTABLE  thrpt    5    54.728 ±   0.997  ops/ms
SearchBenchmark.indexOf          HEAP     WORST_CASE  thrpt    5     1.116 ±   0.032  ops/ms
SearchBenchmark.indexOf     COMPOSITE    RANDOM_256B  thrpt    5   272.040 ±   8.408  ops/ms
SearchBenchmark.indexOf     COMPOSITE     RANDOM_2KB  thrpt    5    44.044 ±   1.936  ops/ms
SearchBenchmark.indexOf     COMPOSITE    PREDICTABLE  thrpt    5    37.634 ±   0.741  ops/ms
SearchBenchmark.indexOf     COMPOSITE  UNPREDICTABLE  thrpt    5    37.623 ±   2.192  ops/ms
SearchBenchmark.indexOf     COMPOSITE     WORST_CASE  thrpt    5     0.766 ±   0.052  ops/ms
SearchBenchmark.indexOf        DIRECT    RANDOM_256B  thrpt    5  1758.146 ± 657.852  ops/ms
SearchBenchmark.indexOf        DIRECT     RANDOM_2KB  thrpt    5   234.797 ±  27.832  ops/ms
SearchBenchmark.indexOf        DIRECT    PREDICTABLE  thrpt    5    49.016 ±  10.052  ops/ms
SearchBenchmark.indexOf        DIRECT  UNPREDICTABLE  thrpt    5    78.379 ±   8.351  ops/ms
SearchBenchmark.indexOf        DIRECT     WORST_CASE  thrpt    5     1.123 ±   0.851  ops/ms

 PSYoungGen      total 76288K, used 48075K [0x000000076ab00000, 0x0000000770000000, 0x00000007c0000000)
  eden space 65536K, 66% used [0x000000076ab00000,0x000000076d5a8fa8,0x000000076eb00000)
  from space 10752K, 40% used [0x000000076eb00000,0x000000076ef49c98,0x000000076f580000)
  to   space 10752K, 0% used [0x000000076f580000,0x000000076f580000,0x0000000770000000)
 ParOldGen       total 175104K, used 16K [0x00000006c0000000, 0x00000006cab00000, 0x000000076ab00000)
  object space 175104K, 0% used [0x00000006c0000000,0x00000006c0004000,0x00000006cab00000)
 Metaspace       used 11909K, capacity 12215K, committed 12544K, reserved 1060864K
  class space    used 1380K, capacity 1469K, committed 1536K, reserved 1048576K
```
Benchmark results (implementation optimized using Kmp algorithm  of ByteBufUtil.indexOf):
```
Benchmark                (bufferType)        (input)   Mode  Cnt     Score    Error   Units
SearchBenchmark.indexOf          HEAP    RANDOM_256B  thrpt    5  2792.422 ± 81.033  ops/ms
SearchBenchmark.indexOf          HEAP     RANDOM_2KB  thrpt    5   300.021 ±  5.835  ops/ms
SearchBenchmark.indexOf          HEAP    PREDICTABLE  thrpt    5   352.242 ± 23.597  ops/ms
SearchBenchmark.indexOf          HEAP  UNPREDICTABLE  thrpt    5   170.998 ±  1.970  ops/ms
SearchBenchmark.indexOf          HEAP     WORST_CASE  thrpt    5    73.004 ±  5.348  ops/ms
SearchBenchmark.indexOf     COMPOSITE    RANDOM_256B  thrpt    5  1061.188 ± 14.019  ops/ms
SearchBenchmark.indexOf     COMPOSITE     RANDOM_2KB  thrpt    5   148.445 ±  3.018  ops/ms
SearchBenchmark.indexOf     COMPOSITE    PREDICTABLE  thrpt    5   135.135 ±  8.789  ops/ms
SearchBenchmark.indexOf     COMPOSITE  UNPREDICTABLE  thrpt    5    75.578 ±  2.820  ops/ms
SearchBenchmark.indexOf     COMPOSITE     WORST_CASE  thrpt    5    41.323 ±  0.144  ops/ms
SearchBenchmark.indexOf        DIRECT    RANDOM_256B  thrpt    5  1880.160 ± 76.520  ops/ms
SearchBenchmark.indexOf        DIRECT     RANDOM_2KB  thrpt    5   204.214 ±  8.540  ops/ms
SearchBenchmark.indexOf        DIRECT    PREDICTABLE  thrpt    5   213.783 ±  9.104  ops/ms
SearchBenchmark.indexOf        DIRECT  UNPREDICTABLE  thrpt    5   131.881 ±  9.242  ops/ms
SearchBenchmark.indexOf        DIRECT     WORST_CASE  thrpt    5    60.850 ±  3.391  ops/ms

Heap
 PSYoungGen      total 76288K, used 48091K [0x000000076ab00000, 0x0000000770000000, 0x00000007c0000000)
  eden space 65536K, 66% used [0x000000076ab00000,0x000000076d5a7e68,0x000000076eb00000)
  from space 10752K, 41% used [0x000000076eb00000,0x000000076ef4ef18,0x000000076f580000)
  to   space 10752K, 0% used [0x000000076f580000,0x000000076f580000,0x0000000770000000)
 ParOldGen       total 175104K, used 16K [0x00000006c0000000, 0x00000006cab00000, 0x000000076ab00000)
  object space 175104K, 0% used [0x00000006c0000000,0x00000006c0004000,0x00000006cab00000)
 Metaspace       used 11911K, capacity 12215K, committed 12544K, reserved 1060864K
  class space    used 1379K, capacity 1469K, committed 1536K, reserved 1048576K
```
Although the improvement of directByteBuf is not large, significant performance improvements have been seen in both headByteBuf and compositeByteBuf, so this optimization is effective

**Comment by franz1981** [comment]
Thanks for the or!
The kmp algorithm is already available in Netty by using some specific byte processor and the stealth allocation of the index using kmp is not welcome IMO, given that the version using byte processor allow to cache the jump table.


I would suggest instead, to use the new SWAR index of (byte) to speed up this index of while leaving the allocation behaviour as it is.

**Comment by skyguard1** [comment]
@franz1981, Thanks for your suggestion,I have seen https://github.com/netty/netty/pull/9955 and its implementation, but I think it is still necessary to implement it on `ByteBufUtil.indexOf()`. Here are the reasons:
1. When using `KmpSearchProcessorFactory`, you need to create `KmpSearchProcessorFactory` through `AbstractSearchProcessorFactory.newKmpSearchProcessorFactory()` first, and users need to call the `ByteBuf.getBytes()` method to pass in a byte array, which is a bit complicated for users
2. The user needs to understand each string matching algorithm and implementation in order to create the corresponding SearchProcessor, but this is not a good choice for the user (just a personal opinion), and use `ByteBufUtil.indexOf()` for comparison is much simpler
3. Caching the next table can improve performance. However, in actual use, the probability of multiple comparisons of the same needle is not certain (just a personal opinion). In fact, the number of times the user actually compares the same byteBuf multiple times Not necessarily as much as we think
These are just my personal opinions, can you explain the optimization points further, thanks

**Comment by franz1981** [comment]
Reading your comments I see that the existing searching API is not simple enough for users and I would rather suggest to improve it: indexOf method should obey the https://en.m.wikipedia.org/wiki/Principle_of_least_astonishment and, although is personal, I don't think allocation (O(n)) while searching to be the expected behaviour here, but as said, maybe it's just me and other users are perfectly fine with it.

Regarding the optimization of the index of byte (SWAR), it's not performing a linear search byte by byte hence could be used to start the needle search  just to fail fast.

**Comment by skyguard1** [comment]
@franz1981, thanks for your suggestion, @chrisvest, Do you think this does not meet the Principle of least astonishment? If it is, I will revert these codes, thanks

**Comment by chrisvest** [comment]
Using `firstIndexOf` (SWAR) might fall flat if the first byte turns out to be very common in the inputs.

I don't think it's a huge problem that `indexOf` allocates. At least, I think most usages of it are limited in how many times they call it. On the other hand, we also have people with very high throughputs, who wish to keep allocations to a minimum because of the non-local effects on response time and stability.

The Two-Way string matching algorithm has similar performance characteristics to KMP and BM, but does not require any table allocation. Perhaps that will be a better choice for our `indexOf` method?
https://en.wikipedia.org/wiki/Two-way_string-matching_algorithm
http://www-igm.univ-mlv.fr/~mac/Articles-PDF/CP-1991-jacm.pdf

**Comment by franz1981** [comment]
@chrisvest as one of such performance critical allocation break users I can say that's an sensible answer :)
Thanks for sharing https://en.wikipedia.org/wiki/Two-way_string-matching_algorithm I didn't know about it, it's even used in glibc!

**Comment by skyguard1** [comment]
Use the two way algorithm to optimize the ByteBufUtil.indexOf() method

Benchmark results (original implementation of ByteBufUtil.indexOf):
```
Benchmark                (bufferType)        (input)   Mode  Cnt     Score     Error   Units
SearchBenchmark.indexOf          HEAP    RANDOM_256B  thrpt    5  1199.046 ± 255.956  ops/ms
SearchBenchmark.indexOf          HEAP     RANDOM_2KB  thrpt    5   156.356 ±  85.901  ops/ms
SearchBenchmark.indexOf          HEAP    PREDICTABLE  thrpt    5    35.895 ±  27.980  ops/ms
SearchBenchmark.indexOf          HEAP  UNPREDICTABLE  thrpt    5    52.720 ±  15.137  ops/ms
SearchBenchmark.indexOf          HEAP     WORST_CASE  thrpt    5     1.083 ±   0.319  ops/ms
SearchBenchmark.indexOf     COMPOSITE    RANDOM_256B  thrpt    5   214.643 ± 285.077  ops/ms
SearchBenchmark.indexOf     COMPOSITE     RANDOM_2KB  thrpt    5    34.849 ±  40.074  ops/ms
SearchBenchmark.indexOf     COMPOSITE    PREDICTABLE  thrpt    5    22.774 ±  24.180  ops/ms
SearchBenchmark.indexOf     COMPOSITE  UNPREDICTABLE  thrpt    5    37.773 ±   5.441  ops/ms
SearchBenchmark.indexOf     COMPOSITE     WORST_CASE  thrpt    5     0.680 ±   0.032  ops/ms
SearchBenchmark.indexOf        DIRECT    RANDOM_256B  thrpt    5  1918.517 ± 173.048  ops/ms
SearchBenchmark.indexOf        DIRECT     RANDOM_2KB  thrpt    5   250.167 ±  21.105  ops/ms
SearchBenchmark.indexOf        DIRECT    PREDICTABLE  thrpt    5    52.147 ±   4.849  ops/ms
SearchBenchmark.indexOf        DIRECT  UNPREDICTABLE  thrpt    5    81.570 ±  12.366  ops/ms
SearchBenchmark.indexOf        DIRECT     WORST_CASE  thrpt    5     1.231 ±   0.815  ops/ms


Heap
 PSYoungGen      total 76288K, used 48247K [0x000000076ab00000, 0x0000000770000000, 0x00000007c0000000)
  eden space 65536K, 66% used [0x000000076ab00000,0x000000076d5c3ed8,0x000000076eb00000)
  from space 10752K, 41% used [0x000000076eb00000,0x000000076ef5a008,0x000000076f580000)
  to   space 10752K, 0% used [0x000000076f580000,0x000000076f580000,0x0000000770000000)
 ParOldGen       total 175104K, used 16K [0x00000006c0000000, 0x00000006cab00000, 0x000000076ab00000)
  object space 175104K, 0% used [0x00000006c0000000,0x00000006c0004000,0x00000006cab00000)
 Metaspace       used 11895K, capacity 12215K, committed 12544K, reserved 1060864K
  class space    used 1379K, capacity 1469K, committed 1536K, reserved 1048576K
```
Benchmark results (implementation optimized using two way algorithm of ByteBufUtil.indexOf):
```
Benchmark                (bufferType)        (input)   Mode  Cnt     Score     Error   Units
SearchBenchmark.indexOf          HEAP    RANDOM_256B  thrpt    5  2973.669 ± 222.865  ops/ms
SearchBenchmark.indexOf          HEAP     RANDOM_2KB  thrpt    5  2998.931 ± 338.832  ops/ms
SearchBenchmark.indexOf          HEAP    PREDICTABLE  thrpt    5   230.671 ±   6.105  ops/ms
SearchBenchmark.indexOf          HEAP  UNPREDICTABLE  thrpt    5   346.158 ±   9.007  ops/ms
SearchBenchmark.indexOf          HEAP     WORST_CASE  thrpt    5   167.851 ±  11.456  ops/ms
SearchBenchmark.indexOf     COMPOSITE    RANDOM_256B  thrpt    5  2417.317 ± 108.826  ops/ms
SearchBenchmark.indexOf     COMPOSITE     RANDOM_2KB  thrpt    5  1981.902 ± 110.820  ops/ms
SearchBenchmark.indexOf     COMPOSITE    PREDICTABLE  thrpt    5   165.867 ±   3.762  ops/ms
SearchBenchmark.indexOf     COMPOSITE  UNPREDICTABLE  thrpt    5   152.748 ±  11.547  ops/ms
SearchBenchmark.indexOf     COMPOSITE     WORST_CASE  thrpt    5    99.890 ±   6.332  ops/ms
SearchBenchmark.indexOf        DIRECT    RANDOM_256B  thrpt    5  2323.481 ±  90.986  ops/ms
SearchBenchmark.indexOf        DIRECT     RANDOM_2KB  thrpt    5  2295.211 ±  36.655  ops/ms
SearchBenchmark.indexOf        DIRECT    PREDICTABLE  thrpt    5   146.229 ±   3.107  ops/ms
SearchBenchmark.indexOf        DIRECT  UNPREDICTABLE  thrpt    5   197.068 ±  11.054  ops/ms
SearchBenchmark.indexOf        DIRECT     WORST_CASE  thrpt    5   152.759 ±  31.891  ops/ms

Heap
 PSYoungGen      total 76288K, used 48702K [0x000000076ab00000, 0x0000000770000000, 0x00000007c0000000)
  eden space 65536K, 67% used [0x000000076ab00000,0x000000076d65daa8,0x000000076eb00000)
  from space 10752K, 39% used [0x000000076eb00000,0x000000076ef32008,0x000000076f580000)
  to   space 10752K, 0% used [0x000000076f580000,0x000000076f580000,0x0000000770000000)
 ParOldGen       total 175104K, used 16K [0x00000006c0000000, 0x00000006cab00000, 0x000000076ab00000)
  object space 175104K, 0% used [0x00000006c0000000,0x00000006c0004000,0x00000006cab00000)
 Metaspace       used 11903K, capacity 12215K, committed 12544K, reserved 1060864K
  class space    used 1378K, capacity 1469K, committed 1536K, reserved 1048576K
```
We can see a great improvement in performance, so this optimization is effective

**Comment by franz1981** [comment]
Happy about it well done!

Just few things I am curious about:
- what happened if the needle is larger of the haystack?
- why the predictable case is slower then the unpredictable one?

For this last question I am going to review the results and the bench: please print the results with turbo boost disabled, fixed proc frequency and using "perfnorm" profiler

**Comment by skyguard1** [comment]
@franz1981, When needle.readableBytes()> haystack.readableBytes(), it will directly return -1, thanks for the suggestion

**Comment by franz1981** [comment]
And what about 
"when the remaining needle to match is larger then the available haystack?"

This can happen if the first N matches on haystack are negative and there are just M < needle::readable bytes left. It should fail fast but I am not sure the original algorithm account for it

**Comment by skyguard1** [comment]
According to my understanding of this algorithm, there is no optimization operation that supports this situation

**Comment by chrisvest** [comment]
Looks like Two Way is doing even better than KMP? That's awesome!

**Comment by franz1981** [comment]
@skyguard1 

> According to my understanding of this algorithm, there is no optimization operation that supports this situation

Yep, I know, but failing fast although won't impact bit O analysis still impact real use cases, especially with "large" needles: I suggest to try modify the algorithm to do it. I've done something similar for the kmp variation we use on https://github.com/franz1981/activemq-artemis/blob/388809a583267de4e0599a4e1e2c76317029eb45/artemis-commons/src/main/java/org/apache/activemq/artemis/utils/algo/KMPNeedle.java#L78 and it proved to be effective in many real use cases.

> Looks like Two Way is doing even better than KMP? That's awesome!

That's why I've suggested to look at the `perfnorm` data: just to look at the branch mispredict and cache friendliness of both (thanks to the reduced allocation and that we just look at 2 different locations, although in different parts, the cache is better used): I believe that if we stress the bench with the right data the perf should confirm the known algorithm complexities.

**Comment by skyguard1** [comment]
A perfnorm  profiler  benchmarking was performed using another linux machine (the previous one was benchmarking on a mac),and the performance is quite different from the previous results
Benchmark results (original implementation using perfnorm  profiler of ByteBufUtil.indexOf):
```
Benchmark                (bufferType)        (input)   Mode  Cnt     Score    Error   Units
SearchBenchmark.indexOf          HEAP    RANDOM_256B  thrpt    5  1379.846 ± 12.699  ops/ms
SearchBenchmark.indexOf          HEAP     RANDOM_2KB  thrpt    5  1377.882 ± 38.710  ops/ms
SearchBenchmark.indexOf          HEAP    PREDICTABLE  thrpt    5   117.111 ±  1.185  ops/ms
SearchBenchmark.indexOf          HEAP  UNPREDICTABLE  thrpt    5   152.722 ±  2.715  ops/ms
SearchBenchmark.indexOf          HEAP     WORST_CASE  thrpt    5    97.151 ±  0.374  ops/ms
SearchBenchmark.indexOf     COMPOSITE    RANDOM_256B  thrpt    5   695.742 ±  8.265  ops/ms
SearchBenchmark.indexOf     COMPOSITE     RANDOM_2KB  thrpt    5  1015.208 ±  5.850  ops/ms
SearchBenchmark.indexOf     COMPOSITE    PREDICTABLE  thrpt    5    65.776 ±  1.209  ops/ms
SearchBenchmark.indexOf     COMPOSITE  UNPREDICTABLE  thrpt    5    76.343 ±  1.915  ops/ms
SearchBenchmark.indexOf     COMPOSITE     WORST_CASE  thrpt    5    53.451 ±  1.036  ops/ms
SearchBenchmark.indexOf        DIRECT    RANDOM_256B  thrpt    5  1163.367 ±  4.682  ops/ms
SearchBenchmark.indexOf        DIRECT     RANDOM_2KB  thrpt    5  1161.318 ± 12.828  ops/ms
SearchBenchmark.indexOf        DIRECT    PREDICTABLE  thrpt    5    84.898 ±  1.604  ops/ms
SearchBenchmark.indexOf        DIRECT  UNPREDICTABLE  thrpt    5    96.589 ±  2.226  ops/ms
SearchBenchmark.indexOf        DIRECT     WORST_CASE  thrpt    5    86.055 ±  0.972  ops/m

```
Benchmark results (two way algorithm implementation using perfnorm  profiler of ByteBufUtil.indexOf):
Benchmark                (bufferType)        (input)   Mode  Cnt     Score    Error   Units
```
SearchBenchmark.indexOf          HEAP    RANDOM_256B  thrpt    5  1383.818 ±  1.111  ops/ms
SearchBenchmark.indexOf          HEAP     RANDOM_2KB  thrpt    5  1383.408 ±  3.494  ops/ms
SearchBenchmark.indexOf          HEAP    PREDICTABLE  thrpt    5   117.058 ±  1.205  ops/ms
SearchBenchmark.indexOf          HEAP  UNPREDICTABLE  thrpt    5   152.435 ±  5.110  ops/ms
SearchBenchmark.indexOf          HEAP     WORST_CASE  thrpt    5    96.976 ±  1.390  ops/ms
SearchBenchmark.indexOf     COMPOSITE    RANDOM_256B  thrpt    5  1245.356 ± 10.715  ops/ms
SearchBenchmark.indexOf     COMPOSITE     RANDOM_2KB  thrpt    5  1009.645 ± 22.645  ops/ms
SearchBenchmark.indexOf     COMPOSITE    PREDICTABLE  thrpt    5    66.048 ±  0.347  ops/ms
SearchBenchmark.indexOf     COMPOSITE  UNPREDICTABLE  thrpt    5    39.037 ±  0.646  ops/ms
SearchBenchmark.indexOf     COMPOSITE     WORST_CASE  thrpt    5    53.392 ±  0.710  ops/ms
SearchBenchmark.indexOf        DIRECT    RANDOM_256B  thrpt    5  1165.479 ±  7.321  ops/ms
SearchBenchmark.indexOf        DIRECT     RANDOM_2KB  thrpt    5  1163.232 ±  7.383  ops/ms
SearchBenchmark.indexOf        DIRECT    PREDICTABLE  thrpt    5    85.377 ±  0.288  ops/ms
SearchBenchmark.indexOf        DIRECT  UNPREDICTABLE  thrpt    5    96.593 ±  0.735  ops/ms
SearchBenchmark.indexOf        DIRECT     WORST_CASE  thrpt    5    86.399 ±  0.938  ops/ms

```
It seems that the performance of the algorithm under perfnorm profiler is completely different

**Comment by franz1981** [comment]
@skyguard1 I expect `-perf perfnorm` to produce many numbers for each benchmarks related cache misses, cpy cycles, mispredictions...where it is?

**Comment by richardstartin** [comment]
> Using firstIndexOf (SWAR) might fall flat if the first byte turns out to be very common in the inputs.

In my experience this is correct - e.g. these are the average search times for text drawn from different distributions (markov generated English and German with transition probabilities inferred from bibles in each language, uniformly random text) for a search implementation which uses SWAR to find the start of the needle before commencing a search using the shift-or algorithm. The distribution of the data has a much larger effect than even the length of the needle (`termlength`).
image

**Comment by skyguard1** [comment]
Sorry, I re-run the benchmark, and the performance of the two way is indeed improved can be seen in the result, but I only have the linux virtual machine, and linux `perf` does not support the detection of cpu cache and other information

**Comment by franz1981** [comment]
@skyguard1 numbers look pretty weird...but it's a bare metal machine or is a virtual env eg AWS?

**Comment by skyguard1** [comment]
It runs on our own cloud platform. I also have a virtual machine on bare metal machine, but perf does not support reading cpu cache information., sorry but I currently don't have a physical machine's linux

**Comment by franz1981** [comment]
So @skyguard1 so in this case better to not trust any of these numbers...
The change is ok per se, considering that it's better then the original implementation so I am already happy with that; I just want to check how it behave on real machines. 
I will try to run it on some physical hw by the next week, before going on vacation.

**Comment by skyguard1** [comment]
@franz1981, great!I am looking forward to seeing the results!thanks for your suggestion

File: buffer/src/test/java/io/netty/buffer/ByteBufUtilTest.java
Comment (chrisvest) [inline buffer/src/test/java/io/netty/buffer/ByteBufUtilTest.java:130]: Remove the `Charsets` import and use `CharsetUtil` everywhere instead.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:232]: Add a comment somewhere that says this is the Two Way algorithm, so those who are unfamiliar with it can learn more if they come across this code and get confused.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:252]: I don't understand what this branch is for, because `per` is always zero, so we are just comparing the `needle` to itself for the first `ell + 1` bytes it contains.

Sounds like this would always be true, and indeed the `else` clause has no test coverage.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:279]: This is the only place where `per` gets a non-zero value.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:321]: Please add some tests that exercises this branch, where `a` and `b` are equal.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:248]: Please replace the `i` and `j` variables in the `max` call with the `maxSuf` calls.

As it is right now, the `i` and `j` variables are used for two different purposes, which is confusing to read. We can instead declare them closer to where they are really used. That would make the code look more like Java and less like C.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java
Comment (chrisvest) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:244]: If `m == 1` we could delegate to `firstIndexOf`.

**Review by chrisvest — CHANGES_REQUESTED** [review]
Thanks for taking the time to work on this.

I had a few comments.

**Comment by chrisvest** [comment]
@skyguard1 One more thing I just noticed: the `indexOf` is meant to operate on the readable bytes, but it doesn't take the `readerIndex` into account. There are also no tests with non-zero `readerIndex`.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java
Comment (trustin) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:232]: and maybe add a link to the page that explains the algorithm nicely. There's one at wikipedia, but it seems a little bit dry - https://en.wikipedia.org/wiki/Two-way_string-matching_algorithm so please feel free to use a better one.

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java
Comment (skyguard1) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:247]: I've done it

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java
Comment (skyguard1) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:333]: I added a test case, when the first byte of needle is equal to the second byte, a will be equal to b

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java
Comment (skyguard1) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java]: Like the memcmp function of the C language, this judgment is used to compare the first n bytes of the storage area str1 and the storage area str2. If needle does not have a repeated prefix, it will enter the else branch, in fact the possibility of entering the else branch will be bigger

Comment (chrisvest) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java]: This will not check for repeated prefixes, though:

```
if (equals(needle, per, needle, per, per + ell + 1)) {
```

It just checks if the needle is equal to itself in the `per` to `per + per + ell + 1` region, which will either throw an `IndexOutOfBoundsException` or be tautologically true. (Notice that the last argument is a length, not an offset)

Comment (skyguard1) [reply buffer/src/main/java/io/netty/buffer/ByteBufUtil.java]: Like the implementation of http://www-igm.univ-mlv.fr/~lecroq/string/node26.html, use an array of length 2 to return the offset value of the `maxSuf()` method, which does not increase the algorithm Space complexity, but  can use the comparison result of `maxSuf()` method

**Comment by skyguard1** [comment]
> @skyguard1 One more thing I just noticed: the `indexOf` is meant to operate on the readable bytes, but it doesn't take the `readerIndex` into account. There are also no tests with non-zero `readerIndex`.

The case where the readerIndex of needle and haystack is not 0 has been considered

**Review by chrisvest — APPROVED** [review]
I think this looks good, now. Thanks!

**Comment by skyguard1** [comment]
@normanmaurer PTAL, please see if there is anything that needs to be modified here, thanks

**Comment by chrisvest** [comment]
A small thing: The two ints returned from `maxSuf` _could_ be returned as one packed `long` value, to guarantee no allocation takes place.

**Comment by skyguard1** [comment]
@chrisvest ,I've done it, thanks

**Comment by skyguard1** [comment]
@chrisvest, thanks!

File: buffer/src/test/java/io/netty/buffer/ByteBufUtilTest.java
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/ByteBufUtilTest.java:141]:
```suggestion
        assertEquals(1, ByteBufUtil.indexOf(needle, haystack));
        haystack.release();
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]
One release is missing … otherwise LGTM

**Comment by skyguard1** [comment]
@normanmaurer, thanks

**Review by normanmaurer — APPROVED** [review]

File: buffer/src/main/java/io/netty/buffer/ByteBufUtil.java
Comment (trustin) [inline buffer/src/main/java/io/netty/buffer/ByteBufUtil.java:232]:
```suggestion
     * This method uses the <a href="https://en.wikipedia.org/wiki/Two-way_string-matching_algorithm">Two-Way
     * string matching algorithm</a>, which yields O(1) space complexity and excellent performance.
```

**Review by trustin — APPROVED** [review]

**Comment by skyguard1** [comment]
@trustin, thanks

**Comment by normanmaurer** [comment]
Once the build has passed I will merge this one

**Comment by chrisvest** [comment]
@skyguard1 Thanks!

**Comment by skyguard1** [comment]
@chrisvest It is my pleasure

**Comment by tananaev** [comment]
@skyguard1 I think there's a bug in the new implementation. As far as I understand, it returns index relative to reader index instead of absolute index, like before. I see that @chrisvest actually suggested that it might be an issue.

**Pull request merged**
